### DISCUSSION TITLE
Can we create variables from response and use those variables for the next request?

### DISCUSSION THREAD

**Discussion by deostroll** [Huachao/vscode-restclient#1325] [Q&A]
For e.g. suppose I login and receive a json response with the accesstoken (at some path, say `$.id`). Can I store this into a variable, and, issue the subsequent requests with the same: `http://localhost:3000/foo/bar?accessToken={{$accessToken}}` ?

**Comment by ropinheiro** [comment]
Yes, you can. Just don't put the $ before accessToken, that is only used for system variables.

**Comment by wilsonneto-swe** [comment]
As in the example, you need to use `@name_of_the_var = {{ firstRequestName.response.body.$.field }}` , so in the next request, you will be able to use it on the body to headers just using `{{ name_of_the_var  }}`.

```
@baseUrl = https://example.com/api

# @name login
POST {{baseUrl}}/api/login HTTP/1.1
Content-Type: application/x-www-form-urlencoded

name=foo&password=bar

###

@authToken = {{login.response.headers.X-AuthToken}}

# @name createComment
POST {{baseUrl}}/comments HTTP/1.1
Authorization: {{authToken}}
Content-Type: application/json

{
    "content": "fake content"
}

###

@commentId = {{createComment.response.body.$.id}}
```

**Comment by jantoney** [comment]
2 topic reply.

## First - Question
Can you save a reply (auth token) to somewhere so it can be accessed by multiple .http files?
I've got many API requests in this project and would like to split it up among a few .http files.
When I follow the below code I've used, there doesn't seem to be a way to save it out.

I guess bonus question would be, it would be great to have a file like a .env that I can place into a .gitignore that I can put a user/pass into that the REST client can read to get authentication and that isn't synced to github like the vscode settings file.

## Second - My example code
To help others, you can mash together a few examples from the readme page to get a good auth experience, like below.

/.vscode/settings.json
```json
{
    "rest-client.environmentVariables": {
      "$shared": {},
      "local": {
        "apiUrl": "http://localhost:8000/api"
      },
      "production": {
        "apiUrl": "https://production.my.url.example.com/api"
      },
      "staging": {
        "apiUrl": "https://staging.my.url.example.com/api"
      },
      "demo": {
        "apiUrl": "https://demo.my.url.example.com/api"
      }
  }
}
```

This will then
- prompt the user for their email and password (so it's not saved in the settings file that is synced to gihub)
- login
- set the variable **authToken** to the auth token that is received in the header. Because my auth header is a bearer token, and the reply includes the word bearer in it, then I don't put it into the next request
- You can then run the next request and it will use the authToken from the login request.

Notes
- You need to run the login request first
- Then run the dashboard stats one (or whatever is for you)
- You must have all your endpoints in the same file as the login command, or just put the login command at the top of each file


my_api.http
```http
@contentType = application/json

###
# @prompt email
# @prompt password
# @name login
POST {{apiUrl}}/auth/login
Content-Type: {{contentType}}

{
  "email": "{{email}}",
  "password": "{{password}}"
}


###
@authToken = {{login.response.headers.Authorization}}


### Test Login by getting dashboard stats
GET {{apiUrl}}/dashboard-stats
Authorization: {{authToken}}
```

**Reply by MominRaza** [reply]
Is it possible to create a shared variable from login response? so we can use it in different files.

**Comment by MominRaza** [comment]
Related update: https://github.com/Huachao/vscode-restclient/pull/1422

This discussion is addressed by the new `@set` directive, which lets you capture response values (`response.headers.*` / `response.body.*`) into shared runtime variables and reuse them across `.http` files.
